ttnn/cpp/ttnn/operations/reduction/topk/device/kernels/compute/topk_local.cpp

```diff
@@ -0,0 +1,172 @@
+    constexpr uint32_t input_dest_start = 0;
+    constexpr uint32_t index_dest_start = 2;
+    constexpr uint32_t input_dest_end = 1;
+    constexpr uint32_t index_dest_end = 3;
+
+    // init pack, compute and unpack
+    ckernel::topk_tile_init();
+    transpose_wh_init(input_cb_index, input_transposed_cb_index);
+    bool ascending = false;
+
+    for(uint32_t ht = 0; ht < Ht; ++ht) {
+        cb_reserve_back(input_transposed_cb_index, Wt);
+        cb_reserve_back(index_transposed_cb_index, Wt);
+
+        // streaming in input and index tiles to transpose and bitonic local sort them, two tiles at a time
+        for (uint32_t wt = 0; wt < Wt; wt+=2) {
+
+            acquire_dst(tt::DstMode::Half);
+            // transpose tiles and then local sort into k groups
+            cb_wait_front(input_cb_index, 2);
+            cb_wait_front(index_cb_index, 2);
+
+            unpack_reconfig_data_format_srca(input_cb_index);
+            transpose_wh_init_short(input_cb_index);
+            transpose_wh_tile(input_cb_index, 0, 0);
+            transpose_wh_tile(input_cb_index, 1, 1);
+
+            unpack_reconfig_data_format_srca(index_cb_index);
+            transpose_wh_init_short(index_cb_index);
+            transpose_wh_tile(index_cb_index, 0, 2);
+            transpose_wh_tile(index_cb_index, 1, 3);
+
+            // llk_topk_sort -> inplace
+            ckernel::topk_local_sort(0, (int) ascending, logk - 1);
+
+            // pack value tiles into cb_intermed0
+            pack_reconfig_data_format(input_transposed_cb_index);
+            pack_tile(0, input_transposed_cb_index);
+            pack_tile(1, input_transposed_cb_index);
+
+            // pack index tiles into cb_intermed1
+            pack_reconfig_data_format(index_transposed_cb_index);
+            pack_tile(2, index_transposed_cb_index);
+            pack_tile(3, index_transposed_cb_index);
+
+            cb_pop_front(input_cb_index, 2);
+            cb_pop_front(index_cb_index, 2);
+            release_dst(tt::DstMode::Half);
+        }
+
+        cb_push_back(input_transposed_cb_index, Wt);
+        cb_push_back(index_transposed_cb_index, Wt);
+
+        // iterative divide and conquer on pairs of tiles (bitonic topk merge and rebuild)
+        // first iteration we compare 0th and 1st tile, then 2nd and 3rd, etc. We get the sorted top 32 values in each pair.
+        // second iteration we compare 0th and 2nd tile, then 4th and 6th, etc.
+        // logWt iteration we compare 0th and Wt/2 tile
+        // single buffer as we can pack tiles back in-place
+        for (uint32_t m_iter = 0; m_iter < logWt; ++m_iter) {
+            bool a = starting_a;
+            cb_wait_front(input_transposed_cb_index, Wt);
+            cb_wait_front(index_transposed_cb_index, Wt);
+
+            for (uint32_t left_ind = 0; left_ind < Wt - (1 << m_iter); left_ind += 2 << m_iter) {
+                uint32_t right_ind = left_ind + (1 << m_iter);
```

Comment:
Can you declare `* << m_iter` as variables and use them? You can save some cycles here

Reply:
Done